Bottom: 6cfcc4de3cb27158eafd35b34ccd045f0d286536
Top:    12aa5d356255440a8b5c150ee53770251e70861a
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-03-12 19:25:20 +0000

Fixup SearchObjectByLocation not to use negative layer numbers


---

diff --git a/src/search.c b/src/search.c
index 5e33b3d..8d7178e 100644
--- a/src/search.c
+++ b/src/search.c
@@ -1129,6 +1129,82 @@ IsPointOnArc (Coord X, Coord Y, Coord Radius, ArcType *Arc)
   return fabs (Distance (X, Y, Arc->X, Arc->Y) - Arc->Width) < Radius + Arc->Thickness / 2;
 }
 
+static unsigned int
+SearchLayerObjectInternal (LayerType *SearchLayer,
+                           LayerType **Result1,
+                           void **Result2,
+                           void **Result3)
+{
+  if (!SearchLayer->On)
+    return NO_TYPE;
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
+      Type & POLYGONPOINT_TYPE &&
+      SearchPointByLocation (locked,
+                             Result1,
+                             (PolygonType **) Result2,
+                             (PointType **) Result3))
+    return (POLYGONPOINT_TYPE);
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
+      Type & LINEPOINT_TYPE &&
+      SearchLinePointByLocation (locked,
+                                 Result1,
+                                 (LineType **) Result2,
+                                 (PointType **) Result3))
+    return (LINEPOINT_TYPE);
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & LINE_TYPE
+      && SearchLineByLocation (locked,
+                               Result1,
+                               (LineType **) Result2,
+                               (LineType **) Result3))
+    return (LINE_TYPE);
+
+    if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
+      Type & ARCPOINT_TYPE &&
+      SearchArcPointByLocation (locked,
+                                Result1,
+                                (ArcType **) Result2,
+                                (PointType **) Result3))
+    return (ARCPOINT_TYPE);
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & ARC_TYPE &&
+      SearchArcByLocation (locked,
+                           Result1,
+                           (ArcType **) Result2,
+                           (ArcType **) Result3))
+    return (ARC_TYPE);
+
+  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & TEXT_TYPE
+      && SearchTextByLocation (locked,
+                               Result1,
+                               (TextType **) Result2,
+                               (TextType **) Result3))
+    return (TEXT_TYPE);
+
+  if (Type & POLYGON_TYPE &&
+      SearchPolygonByLocation (locked,
+                               Result1,
+                               (PolygonType **) Result2,
+                               (PolygonType **) Result3))
+    {
+      if (HigherAvail)
+        {
+          BoxType *box =
+            &(*(PolygonType **) Result2)->BoundingBox;
+          double area =
+            (double) (box->X2 - box->X1) * (double) (box->X2 - box->X1);
+          if (HigherBound < area)
+            break;
+          else
+            return (POLYGON_TYPE);
+        }
+      else
+        return (POLYGON_TYPE);
+    }
+}
+
 /* ---------------------------------------------------------------------------
  * searches for any kind of object or for a set of object types
  * the calling routine passes two pointers to allocated memory for storing
@@ -1154,6 +1230,7 @@ SearchObjectByLocation (unsigned Type,
   double HigherBound = 0;
   int HigherAvail = NO_TYPE;
   int locked = Type & LOCKED_TYPE;
+  unsigned int type;
   /* setup variables used by local functions */
   PosX = X;
   PosY = Y;
@@ -1253,76 +1330,13 @@ SearchObjectByLocation (unsigned Type,
 	  SearchLayer = &PCB->Data->BACKSOLDERMASKLAYER;
 	  if (!PCB->InvisibleObjectsOn)
 	    continue;
-	}
-      if (SearchLayer->On)
-	{
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
-	      Type & POLYGONPOINT_TYPE &&
-	      SearchPointByLocation (locked,
-				     (LayerType **) Result1,
-				     (PolygonType **) Result2,
-				     (PointType **) Result3))
-	    return (POLYGONPOINT_TYPE);
-
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
-	      Type & LINEPOINT_TYPE &&
-	      SearchLinePointByLocation (locked,
-					 (LayerType **) Result1,
-					 (LineType **) Result2,
-					 (PointType **) Result3))
-	    return (LINEPOINT_TYPE);
+        }
 
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & LINE_TYPE
-	      && SearchLineByLocation (locked,
-				       (LayerType **) Result1,
-				       (LineType **) Result2,
-				       (LineType **) Result3))
-	    return (LINE_TYPE);
-
-	    if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 &&
-	      Type & ARCPOINT_TYPE &&
-	      SearchArcPointByLocation (locked,
-					(LayerType **) Result1,
-					(ArcType **) Result2,
-					(PointType **) Result3))
-	    return (ARCPOINT_TYPE);
-
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & ARC_TYPE &&
-	      SearchArcByLocation (locked,
-				   (LayerType **) Result1,
-				   (ArcType **) Result2,
-				   (ArcType **) Result3))
-	    return (ARC_TYPE);
-
-	  if ((HigherAvail & (PIN_TYPE | PAD_TYPE)) == 0 && Type & TEXT_TYPE
-	      && SearchTextByLocation (locked,
-				       (LayerType **) Result1,
-				       (TextType **) Result2,
-				       (TextType **) Result3))
-	    return (TEXT_TYPE);
-
-	  if (Type & POLYGON_TYPE &&
-	      SearchPolygonByLocation (locked,
-				       (LayerType **) Result1,
-				       (PolygonType **) Result2,
-				       (PolygonType **) Result3))
-	    {
-	      if (HigherAvail)
-		{
-		  BoxType *box =
-		    &(*(PolygonType **) Result2)->BoundingBox;
-		  double area =
-		    (double) (box->X2 - box->X1) * (double) (box->X2 - box->X1);
-		  if (HigherBound < area)
-		    break;
-		  else
-		    return (POLYGON_TYPE);
-		}
-	      else
-		return (POLYGON_TYPE);
-	    }
-	}
+      type = SearchLayerObjectInternal (SearchLayer, Result1, Result2, Result3);
+      if (type != NO_TYPE)
+        return type;
     }
+
   /* return any previously found objects */
   if (HigherAvail & PIN_TYPE)
     {
